Revival of "Allows AIs to deploy to borgs" by Shadowlight213

GunHog opened the pull request:
This PR is a continuation of https://github.com/tgstation/tgstation/pull/23571 by @Shadowlight213 

:cl: Gun Hog and Shadowlight213
add: The AI may now deploy to cyborgs prepared as AI shells. The module to do this may be research in the exosuit fabricator. Simply slot the module into a completed cyborg frame as with an MMI, or into a playerless (with no ckey) cyborg.
add: AI shells and AIs controlling a shell can be determined through the Diagnostic HUD.
add: AIs can deploy to a shell using the new action buttons or by simply clicking on it.
experimental: An AI shell will always have the laws of its controlling AI.
/:cl:

Some PR notes:
- This code is a modification of the PR mentioned above. The concept remains the same.
- The code needs adapting to make better use of RR's Soullink system.
- A diagnostic HUD entry has been added to track remote AI control of a shell for both mobs involved.
- AI Shells are created using the B.O.R.I.S. module, as so named by three commentators on the previous PR.
- The B.O.R.I.S. module can turn cyborgs without a player key into a shell, or it can spawn a new cyborg shell by using the item on a completed cyborg frame.
- The AI has a count of all AI Shells on the map. This does not account for shells taken by other AIs, however.
- Damage, power failure, and EMP events force the AI back to its core.
- The AI must choose a shell from a list of all available AI shells the first time it uses the feature.
- The AI player gets action buttons for connecting to a new AI shell, returning to its core from the shell, and returning to the last shell it used.
- Whenever the AI transfers to a shell or back to its core, its lawset will be displayed, in the event that they have changed.
- The AI may click on an AI shell to deploy to it.

**Notable design decisions:**
- **EDIT:** The Shell will now always have the laws of the AI.
- Clockwork Cult conversion of an AI shell also converts the AI player at its core (carried over from the first PR). 

Since Shadowlight's PR was so well received, I thought I could continue it after he took a break from coding. 


![screenshot 2017-03-17 14 13 46](https://cloud.githubusercontent.com/assets/4955510/24063618/83347a06-0b2e-11e7-81a1-b076d53e1a93.png)

Commit:
Premerge commit

Commit:
Merge branch 'master' into AI_Deploy

# Conflicts:
#	code/modules/mob/living/silicon/robot/robot.dm

Commit:
Phase 2

- Add Diag HUD entries
- Modify upgrade code to make a new shell
- Allow B.O.R.I.S. to work on keyless borg mobs
- AI tracking of shell count
- Damage, EMP, and power failure force AI back to core
- Show deployed status on AI
- Update law flavor text for connecting to borg as AI
- Generalize creating AI shells
- Add B.O.R.I.S. to mechfab
- Add action buttons

Commit:
Merge remote-tracking branch 'remotes/upstream/master' into AI_Deploy

# Conflicts:
#	code/modules/mob/living/silicon/robot/robot.dm

Commit:
Remove unused arg

- Removed an unused argument in Initialize()

coiax commented:
Can the AI go full slime person and have three hundred bodies around the station?

lzimann commented on code/datums/antagonists/datum_clockcult.dm:
```diff
@@ -18,6 +18,14 @@
+			var/mob/living/silicon/ai/AI = R.mainframe
+			R.undeploy()
+			var/converted = add_servant_of_ratvar(AI, silent_update)
+			AI << "<span class='userdanger'>Anomaly Detected. Returned to core!</span>"	//The AI needs to be in its core to properly be converted
```
`to_chat(AI, msg)` please

GunHog replied:
Apologies, I forgot to change that portion of the code. I will catch it in my next commit.

GunHog commented:
> Can the AI go full slime person and have three hundred bodies around the station?

Certainly, if Robotics is so autistic as to make several shells for you!

lzimann commented on code/modules/mob/living/silicon/ai/ai.dm:
```diff
@@ -907,10 +914,62 @@ var/list/ai_list = list()
+		var/mob/living/silicon/robot/R = borgie
+		if(R.shell && !R.deployed && (R.stat != DEAD) && (!R.connected_ai ||(R.connected_ai == src)))
+			possible += R
+			world << "Possible:[R], [possible.len]"
```
debug message

GunHog replied:
Dag nabbit! I will nip this one on next commit as well!

lzimann commented on code/modules/mob/living/silicon/robot/robot.dm:
```diff
@@ -424,7 +434,9 @@
 	else if(istype(W, /obj/item/weapon/screwdriver) && opened && cell)	// radio
-		if(radio)
+		if(shell)
+			user << "You cannot seem to open the radio compartment"	//Prevent AI radio key theft
```
to_chat

lzimann commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -248,6 +248,41 @@
+		var/obj/item/borg/upgrade/ai/M = W
+		if(check_completion())
+			if(!isturf(loc))
+				user << "<span class='warning'>You cannot install[M], the frame has to be standing on the ground to be perfectly precise!</span>"
```
to_chat

lzimann commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -248,6 +248,41 @@
+				user << "<span class='warning'>You cannot install[M], the frame has to be standing on the ground to be perfectly precise!</span>"
+				return
+			if(!user.drop_item())
+				user << "<span class='warning'>[M] is stuck to your hand!</span>"
```
to_chat

GunHog replied:
Thanks for catching all of these.

coiax commented:
Will the AI still hear all radio channels while deployed?

Repukan commented:
Make syndicate Borg shells have syndicate laws.

It would be unbalanced and contrarian to have Asimov on those shells.

coiax commented:
There's no such thing as a syndicate borg shell.

Repukan commented:
>Wiz round
>Xenobio
>Xenobio

Repukan commented:
He also mentioned borgs with no player attached to them that can be turned into shells.

Read the PR

GunHog commented:
> Will the AI still hear all radio channels while deployed?

That is the intent! The most recent change I have to test a bit more, to make sure AI keeps all its channels, including Syndie, as a borg.

>Make syndicate Borg shells have syndicate laws.

It does :smile: 
![image](https://cloud.githubusercontent.com/assets/4955510/24022670/ebab11b2-0a75-11e7-8d70-f5802e03ffff.png)

Repukan commented:
If anything AI should lose all channels and keep only common. This can be used as a free "get out of a dangerous" situation card, no?

An AI is an advanced program. If you think about it, this is cramming high stress software into a toaster. Think playing AAA games in a non optimized computer.

GunHog commented:
Lore/Gameplay wise, it is the AI remotely controlling the cyborg. It is not crammed into the borg (codewise, it is though). The AI preserves its radio channels because it has not really gone anywhere. 

Note that it is no 'get out of a dangerous situation card', as **the core remains where it is, and should the core die, the AI player will lose the borg.** Furthermore, if someone is screwing with your core (hitting it, carding it, powering it off), then you are forced back and must deal with the situation.

XDTM commented:
Yay my name made it in

Repukan commented:
That last thing is gay, make it go unnoticed unless the AI player keeps doing check ups on themselves

GunHog commented:
Are you so sure? This way, an AI cannot fight chamber intruders if they reach the core. A cyborg is much better at combat than an AI is.

I cannot change it for carding, because it represents physically pulling the AI out of its core and cutting its wireless connection. The AI may return to the shell if the holder turns wireless back on, though.

EMP and power loss represent a disruption in the control signal. Since the AI is knocked out by both of these, it is only logical it lose its remote connection.

Forced return due to taking damage is up for debate, however. The signal is not lost until the AI is killed in this case. I require additional opinions before I will change this, though.

Repukan commented:
Why would we put an immovable thing in a deep fortress if you're just gonna let it move later?

If your reasoning is that the AI should be able to defend itself as a cyborg, then that opens up a path to nerf the autism fortress they reside in.

GunHog commented:
The current code would **prevent** the AI defending itself as a cyborg should anyone reach the chamber.

coiax commented:
The AI's laws should shadow any lawset of the borg chassis, that would be the method of least surprise.

ExcessiveUseOfCobblestone commented:
what did he mean by this

Repukan commented:
I see, nevermind then, I misunderstood

Commit:
To_chat, remove debug

GunHog commented:
> The AI's laws should shadow any lawset of the borg chassis, that would be the method of least surprise.

Emagging a borg emags the AI, or directly applying a law to the borg also applies it to the AI, you mean? I am worried that would make AI shells a portable AI upload. I am going to need an official decision on that given the implications, but I do welcome your feedback!

coiax commented:
Oh, I didn't consider the possibility of an AI inhabited borg being emagged. It should probably just ignore the emag then, because the borg is just remote controlled, not inhabited by the mind of the AI.

Doppeltongue commented:
If the AI is remote controlling a shell, then it should ignore all laws applied to the shell, or the shell shouldn't have laws so we avoid confusion.

RemieRichards commented:
The shell's laws should always be the AIs, no changing, no emagging, a perfect mirror of the AI's laws.

GunHog commented:
I will make it so in my next commit. Question then, what should happen when someone attempts to emag the shell?

RemieRichards commented:
kick the AI out and break the shell.

GunHog commented:
"break" as in kill the mob or something?

RemieRichards commented:
just remove the shell upgrade?

GunHog commented:
That brings up a case I should test, if someone tries to manually reset a shell, the upgrade would drop. Should both cases result in a mindless, unusable borg mob, or simply deconstruct it on the spot?

coiax commented:
If the upgrade drops, then you have a mindless borg mob.

ExcessiveUseOfCobblestone commented:
honestly, if they are emagged they should be able to be hit with a camera bug which allows YOU to CONTROL the borg.

Shadowlight213 commented:
@Gunhog you should probably see if you can do a eye recolor or something for AI shells so it's more evident that it's a shell instead of a borg.

Anonmare commented:
Maybe you can make it so that an emag kills the shell, kicks the AI out of the shell and (ICly) wipes its memory of the incident.

Also for eye colour I vote either blue, purple or red

Shadowlight213 commented:
Emagging an ai in a occupied shell could maybe give the emagger temporary camera invisibility like the ling power maybe? Emag temporarily scrambles the user's camera signature or something.
It's the only thing I can think of that isn't outweighed by the AI being tipped off about the emagger.
For CC, converting an AI shell could make the entire cult invisible to the AI.

Commit:
Phase 3

- Fixed being able to insert more than one B.O.R.I.S.
- The Deploy To Shell ability is now an action button.
- Changed how AI shells are named -> <Module> AI Shell <number> for
empty ones, and <AI name> shell <number> - <module> for used ones
- Resetting a borg now removes the B.O.R.I.S. like all other upgrades,
disconnecting the AI and leaving a mindless borg. The module must be
re-inserted to make it a shell again.
- Emagging an AI shell now simply resets the borg.
- AI Shells will now always lawsync with their controlling AI,
regardless of circumstance.
- Empty AI Shells no longer show on the cyborg upload computer.
- Sprite changes to the AI shell action buttons

GunHog commented:
@coiax, this is now the case. 
@ExcessiveUseOfCobblestone: That is a feature for another PR! 
@Shadowlight213: It is pretty obvious from the name alone, but if you got me sprites, maybe I can do that. Maybe. 
@Anonmare Emagging the shell resets it, disconnecting the AI. It should look like a normal reset to most unrobust AIs.
@Shadowlight213: Current, ClockCults converting an AI shell with an AI in it will convert the AI, as in your previous PR. You cannot convert a shell when the AI is not using it, as it has no mind.

New in this commit:
 - Fixed being able to insert more than one B.O.R.I.S.
- The Deploy To Shell ability is now an action button.
- Changed how AI shells are named -> <Module> AI Shell <number> for
empty ones, and <AI name> shell <number> - <module> for used ones
- Resetting a borg now removes the B.O.R.I.S. like all other upgrades,
disconnecting the AI and leaving a mindless borg. The module must be
re-inserted to make it a shell again.
- Emagging an AI shell now simply resets the borg.
- AI Shells will now always lawsync with their controlling AI,
regardless of circumstance.
- Empty AI Shells no longer show on the cyborg upload computer.
- Sprite changes to the AI shell action buttons

GunHog commented:
Not currently, but I will add that feature in my next commit.

Commit:
Click-to-deploy, examine, and button fix

- The AI may now deploy to a valid AI shell by clicking on it..
- The AI Shell's examine message now shows when the shell is in use.
- Fixed not getting a message when trying to deploy to a shell, but none
can be taken.

GunHog commented:
New in this commit:
- The AI may now deploy to a valid AI shell by clicking on it. Idea by @coiax 
- The AI Shell's examine message now shows when the shell is in use.
- Fixed not getting a message when trying to deploy to a shell, but none
can be taken.

Commit:
Merge remote-tracking branch 'remotes/upstream/master' into AI_Deploy

# Conflicts:
#	icons/mob/actions.dmi

Commit:
Merge branch 'master' into AI_Deploy

# Conflicts:
#	icons/obj/module.dmi

GunHog commented:
The core of this is done, but I am certainly welcome to suggestions regarding the mechanics and such. There are still more cases I should test, though. _Note that anything that RR or other maintainers say I should have, I cannot change based on player feedback until I have permission._

Anonmare commented:
How will this interact with traitor AIs and their mechanics? Like, will a traitor AI with the camera upgrade still have night vision when deploying to a shell, or still be able to overload/override machinery and hack APCs?

Also, I'd personally like to be suggest being able to buy an AI shell as a traitor AI, at a steep mark-up price of course. It's be pretty handy, especially if you get the robot army objective and no roundstart cyborgs

coiax commented on code/game/objects/items/robot/robot_upgrades.dm:
```diff
@@ -381,3 +381,22 @@
 	return 1
+
+/obj/item/borg/upgrade/ai
+	name = "b.o.r.i.s. module"
```
I think you can call it `B.O.R.I.S.`, lower case isn't used on acronyms. Could also just go BORIS.

GunHog replied:
Oh, thought it was some kind of naming rule against that. Will change in next commit.

Commit:
B.O.R.I.S.

b.o.r.i.s. -> B.O.R.I.S.

GunHog commented:
Sorry, I missed this before.
> How will this interact with traitor AIs and their mechanics? Like, will a traitor AI with the camera upgrade still have night vision when deploying to a shell, or still be able to overload/override machinery and hack APCs?

That is a negatory. That code is entirely AI only. That is too large a change to apply in to this PR.

>Also, I'd personally like to be suggest being able to buy an AI shell as a traitor AI, at a steep mark-up price of course. It's be pretty handy, especially if you get the robot army objective and no roundstart cyborgs

An interesting idea, worth checking into this later if this PR is merged.

Commit:
Further tweaks

- Destroyed (not dead) AI shells are now properly removed from the AI
shells list.
- Prevented use of law boards on AI shells.

GunHog commented:
New in this commit:
- Destroyed (not dead) AI shells are now properly removed from the AI
shells list.
- Prevented use of law boards on AI shells.

RemieRichards commented on code/game/objects/items/robot/robot_upgrades.dm:
```diff
@@ -381,3 +381,22 @@
+
+/obj/item/borg/upgrade/ai
+	name = "B.O.R.I.S. module"
+	desc = "Bluespace Optimized Remote Intelligence Synchronization. An uplink device which takes the place of an MMI in cyborg endskeletons, creating a robotic shell controlled by an AI."
```
endoskeletons

GunHog replied:
Noted!

Commit:
Emag warning + spelling fix

-Fix bad spelling on BORIS description.
- Players attempting to emag an AI shell's cover will now get a warning
about trying to emag its interface.

GunHog commented:
New in this commit:
-Fix bad spelling on BORIS description.
- Players attempting to emag an AI shell's cover will now get a warning
about trying to emag its interface.

The warning is intended to avoid slaying noobs that do not know how AI shells work.

RemieRichards commented on code/modules/mob/living/silicon/robot/robot_defense.dm:
```diff
@@ -94,6 +94,9 @@
 		else
 			to_chat(user, "<span class='warning'>The cover is already unlocked!</span>")
+
+		if(shell) //A warning to Traitors who may not know that emagging AI shells does not slave them.
```
should this not be under the ```if(locked)``` part?
because:

The cover is already unlocked!
Spibble seems to be controlled remotely! Emagging the interface may not work as expected.

seems odd to me.

Commit:
RemRem Commit

- Moves the emag warning to only display when unlocking the borg's door,
not after it is unlocked as well.

GunHog commented:
Done, RR. I am sorry this is taking so long, changes are a PAIN to test with the lighting system taking up to two full minutes to initialize.

RemieRichards commented:
That's fine.

Commit:
wires tweak and edge case fix

- Cutting the AI wire on a deployed shell will disconnect its
controlling AI, allowing any AI to connect to it. (Including the AI that
was just disconnecteD)
- Pulse the AI wire on a deployed shell will also disconnect the
controlling and AI and possibly assign it to a different AI.
- Fixed an edge case where an deployed AI getting a new lawset would
fail to sync with the shell should the shell have law updating disabled.

GunHog commented:
New in this commit:
- Cutting the AI wire on a deployed shell will disconnect its
controlling AI, allowing any AI to connect to it. (Including the AI that
was just disconnecteD)
- Pulse the AI wire on a deployed shell will also disconnect the
controlling and AI and possibly assign it to a different AI.
- Fixed an edge case where an deployed AI getting a new lawset would
fail to sync with the shell should the shell have law updating disabled.

XDTM commented:
Got an alternate acronym, Basic Online Robotic Interface System

GunHog commented:
Get me a better, more AI/computery/cooler replacement for "Basic" and you have a deal!

Commit:
New sprite for B.O.R.I.S.

- New codersprite for BORIS!

GunHog commented:
New B.O.R.I.S. coder sprite:
![boris](https://cloud.githubusercontent.com/assets/4955510/24218991/3fd61b0a-0f13-11e7-9fed-345b30b99315.gif)

XDTM commented:
Binary? I've been looking for a better-flowing one for a while, but no ideas yet.

lzimann commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -248,6 +248,41 @@
+			var/mob/living/silicon/robot/O = new /mob/living/silicon/robot/shell(get_turf(src))
+
+			if(!aisync)
+				lawsync = 0
```
FALSE

lzimann commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -248,6 +248,41 @@
+					O.connected_ai = forced_ai
+				O.notify_ai(4)
+			if(!lawsync)
+				O.lawupdate = 0
```
FALSE

lzimann commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -248,6 +248,41 @@
+			forceMove(O)
+			O.robot_suit = src
+			if(!locomotion)
+				O.lockcharge = 1
```
TRUE

lzimann commented on code/game/objects/items/robot/robot_upgrades.dm:
```diff
@@ -381,3 +381,22 @@
+		to_chat(usr, "<span class='warning'>This unit is already an AI shell!</span>")
+		return
+	if(R.key) //You cannot replace a player unless the key is completely removed.
+		to_chat(usr, "<span class='warning'>Intelligence patterns detected in this [R.braintype]. Aborting.</span>")
```
It's just a minor nitpick of mine, but you could do an early return here so you don't have do make an else down here.

GunHog replied:
Done!

lzimann commented on code/modules/mob/living/silicon/robot/robot.dm:
```diff
@@ -849,6 +866,8 @@
 			to_chat(connected_ai, "<br><br><span class='notice'>NOTICE - Cyborg module change detected: [name] has loaded the [designation] module.</span><br>")
 		if(3) //New Name
 			to_chat(connected_ai, "<br><br><span class='notice'>NOTICE - Cyborg reclassification detected: [oldname] is now designated as [newname].</span><br>")
+		if(4) //New Shell
```
These could all be made a define

lzimann reviewed (changes requested):
Some small things, but overall good.

Commit:
DEFINES!

- Made defines for AI Notifications
- TRUE/FALSE -> 1/0

Commit:
Catch last set

Forgot one set lol

coiax commented on code/__DEFINES/robots.dm:
```diff
@@ -34,3 +34,9 @@
+#define		NEW_BORG		1
+#define		NEW_MODULE	2
+#define		RENAME		3
+#define		AI_SHELL		4
```
thank god

Commit:
Alignment

uses spaces instead of tabs

Mysak0CZ commented:
I found this PR when looking through what i missed while away.
@GunHog 
About Emagging i thing, it would be better, if it worked like standart cyborg emag withount lawchange (only unlocking illegal equipment). It will be better for traitors that already subverted AIs/traitors cooperating with traitor AI.

GunHog commented:
It is pointless for Traitor AIs, as they can already hack borgs via the RD's computer. As for subverted AIs, it makes more sense. My method of simply resetting the shell is probably not the best solution, either. 

This is worth discussion, and an easy change to make. I will have to know what coderbus thinks of it.
